lib/wibox/widget/graph.lua

```diff
@@ -159,6 +159,20 @@ local properties = { "width", "height", "border_color", "stack",
                      "capacity",
 }
 
+local function graph_gather_drawn_values_num_stats(self, new_value)
+    if not (new_value >= 0) then
+        return -- is negative or NaN
+    end
+
+    local last_value = self._private.last_drawn_values_num or 0
+    -- Grow instantly and shrink slow
+    if new_value < last_value then
+        self._private.last_drawn_values_num = last_value - 1
```

Comment:
I'm not *setting* the initial value of `last_drawn_values_num` above 8000, I'm allowing up to 8k elements to accumulate until the first call to `draw()`. The first call to `draw()` will set `last_drawn_values_num` for the first time, after which the number of kept elements will immediately plunge to that + approximately 64-128 elements for safety. Shrinking by halfway steps is too fast and will defeat the purpose, it's `log_2 n` vs `n` difference, e.g. 7 steps vs 128. So the component would fail to tolerate some short temporary size shrinkages.

I would actually almost think that all of this is overengineered, and one could just keep 8k elements at all times, because memory and performance are unlikely to be the problem for 99.9999% of users, while annoyingly missing values are. But well, the current algorithm is also not that complex either and allows for the eventual optimal performance. Either way, shrinking faster is not a priority and the memory would not be reclaimed by Lua after first use anyway.

Reply:
Ah, I see. I missed the connection then.
Thank you for clarifying.

Reply:
Uhm... what if the widget is shown in multiple places at once with multiple sizes...?

Reply:
The quoted lines are supposed to handle exactly the case, by keeping the largest `last_drawn_number_num` and only slowly decreasing it, when widget is drawn with smaller sizes. There are additional 64-128 safety elements kept, so the widget could be drawn in 64 places at once with smaller sizes and the user won't notice the lack of values in the largest instance. If they somehow want more widgets, there's the new `capacity` property, which allows to override the heuristic and keep any number of elements the user wants.